File: kmp-nativecoroutines-core/src/nativeCoroutinesMain/kotlin/com/rickclephas/kmp/nativecoroutines/NativeCallback.kt
package com.rickclephas.kmp.nativecoroutines

/**
 * A callback with a single argument.
 *
 * We don't want the Swift code to know how to get the [Unit] object, so we'll provide it as the second argument.
 * This way Swift can just return the value that it received without knowing what it is/how to get it.
 */
public typealias NativeCallback<T> = (T, NativeUnit) -> NativeUnit

/**
 * Invokes the callback with the specified [value].
 */
internal inline operator fun <T> NativeCallback<T>.invoke(value: T) =
    invoke(value, NativeUnit)

/**
 * A callback with two arguments.
 *
 * We don't want the Swift code to know how to get the [Unit] object, so we'll provide it as the third argument.
 * This way Swift can just return the value that it received without knowing what it is/how to get it.
 */
public typealias NativeCallback2<T1, T2> = (T1, T2, NativeUnit) -> NativeUnit

/**
 * Invokes the callback with the specified [value1] and [value2].
 */
internal inline operator fun <T1, T2> NativeCallback2<T1, T2>.invoke(value1: T1, value2: T2) =
    invoke(value1, value2, NativeUnit)

File: KMPNativeCoroutinesCore/NativeCallback.swift
//
//  NativeCallback.swift
//  KMPNativeCoroutinesCore
//
//  Created by Rick Clephas on 06/06/2021.
//

/// A callback with no arguments.
public typealias NativeCallback = () -> NativeUnit?

/// A callback with a single argument.
public typealias NativeCallback1<T> = (T) -> NativeUnit?

/// A no-op function that has the signature of a `NativeCallback1`.
public func EmptyNativeCallback1<T>(value: T) -> NativeUnit? { nil }

/// A callback with two arguments.
public typealias NativeCallback2<T1, T2> = (T1, T2) -> NativeUnit?

/// A no-op function that has the signature of a `NativeCallback2`.
public func EmptyNativeCallback2<T1, T2>(value1: T1, value2: T2) -> NativeUnit? { nil }
